doc/design_recfuns.md

```diff
@@ -0,0 +1,93 @@
+# Design for handling recursive functions
+
+Main source of inspiration is [Sutter, Köksal & Kuncak 2011],
+as implemented in Leon, but the main
+differences is that we should unroll function definitions directly from the
+inside of Z3, in a backtracking way. Termination and fairness are ensured by
+iterative-deepening on the maximum number of unrollings in a given branch.
+
+## Unfolding
+
+The idea is that every function definition `f(x1…xn) := rhs[x1…xn]` is
+compiled into:
+
+- a list of cases `A_f_i[x1…xn] => f(x1…xn) = rhs_i[x1…xn]`.
+  When `A_f_i[t1…tn]` becomes true in the model, `f(t1…tn)` is said to be
+  *unfolded* and the clause `A_f_i[t1…tn] => f(t1…tn) = rhs_i[t1…tn]`
+  is added as an auxiliary clause.
+- a list of constraints `Γ_f_i[x1…xn] <=> A_f_i[x1…xn]`
+  that states when `A_f_i[x1…xn]` should be true, depending on inputs `x1…xn`.
+  For every term `f(t1…tn)` present in congruence closure, we
+  immediately add all the `Γ_f_i[t1…tn] <=> A_f_i[t1…tn]` as auxiliary clauses
+  (maybe during internalization of `f(t1…tn)`?).
+
+where each `A_f_i[x1…xn]` is a special new predicate representing the
+given case of `f`, and `rhs_i` does not contain any `ite`.
+We assume pattern matching has been compiled to `ite` beforehand.
+
+For example, `fact(n) := if n<2 then 1 else n * fact(n-1)` is compiled into:
+
+- `A_fact_0[n] => fact(n) = 1`
+- `A_fact_1[n] => fact(n) = n * fact(n-1)`
+- `A_fact_0[n] <=> n < 2`
+- `A_fact_1[n] <=> ¬(n < 2)`
+
+The 2 first clauses are only added when `A_fact_0[t]` is true
+(respectively `A_fact_1[t]` is true).
+The 2 other clauses are added as soon as `fact(t)` is internalized.
+
+## Termination
+
+To ensure termination, we define variables:
+
+- `unfold_depth: int`
```

Comment:
we can add configuration options for this along with the other smt parameters: edit smt/params/smt_params_helper.pyg, smt_params.cpp and smt_params.h to add them.
Access them from the theory plugin from get_fparams(), or using the form:
     params_ref p = smt_params_helper p(get_context().get_params());
     p.max_unfound_depth()...